src/Snak/ExtendedPropertyValueSnak.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace Wikibase\DataModel\Snak;
+
+use DataValues\DataValue;
+use InvalidArgumentException;
+use Wikibase\DataModel\Entity\EntityId;
+use Wikibase\DataModel\Entity\PropertyId;
+
+/**
+ * @since 3.1
+ *
+ * @licence GNU GPL v2+
+ * @author Adam Shorland
+ */
+class ExtendedPropertyValueSnak extends PropertyValueSnak {
+
+	/**
+	 * @var DataValue[]
+	 */
+	private $extendedDataValues = array();
+
+	/**
+	 * Support for passing in an EntityId instance that is not a PropertyId instance has
+	 * been deprecated since 0.5.
+	 *
+	 * @param PropertyId|EntityId|integer $propertyId
+	 * @param DataValue $dataValue
+	 * @param DataValue[] $extendedDataValues
+	 */
+	public function __construct( $propertyId, DataValue $dataValue, array $extendedDataValues ) {
+		parent::__construct( $propertyId, $dataValue );
+
+		foreach ( $extendedDataValues as $key => $extraDataValue ) {
+			if ( !$extraDataValue instanceof DataValue || !is_string( $key ) ) {
+				throw new InvalidArgumentException(
+					'$extraDataValues must be an array of DataValue objects with string keys'
+				);
+			}
+		}
+
+		$this->extendedDataValues = $extendedDataValues;
+	}
+
+	/**
+	 * @return DataValue[]
+	 */
+	public function getExtendedDataValues() {
+		return $this->extendedDataValues;
+	}
+
+	/**
+	 * @param string $key
+	 *
+	 * @return DataValue|null
+	 */
+	public function getExtendedDataValue( $key ) {
+		if ( isset( $this->extendedDataValues[$key] ) ) {
+			return $this->extendedDataValues[$key];
+		}
+		return null;
+	}
+
+	/**
+	 * @see Serializable::serialize
+	 * @note This overrides PropertyValueSnak::serialize
```

Comment:
Yes, getArrayToSerialize, or toArray, would be nice. Could be protected.
NOTE: SnakSerializer doesn't use this at all. We will have to make SnakSerializer aware of ExtendedPVS.